### PR TITLE
[spirv] Booleans do not have a physical layout.

### PR DESCRIPTION
**Pull request by ehsannas**
Addresses issue #1136.

### CODE REVIEW

**Comment by ehsannas** [comment]
Looks like i should rebase

File: tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.boolean.hlsl [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// CHECK-NEXT:        [[boolVec:%\d+]] = OpINotEqual %v3bool [[uintVec]] [[v3uint0]]
+// CHECK-NEXT:                           OpStore %e [[boolVec]]
+    bool3   e = frameConstants.boolMat[0];
+
+// CHECK:      [[FrameConstants:%\d+]] = OpAccessChain %_ptr_Uniform_FrameConstants %CONSTANTS %int_0
+// CHECK-NEXT:        [[uintPtr:%\d+]] = OpAccessChain %_ptr_Uniform_uint [[FrameConstants]] %int_2 %uint_1 %uint_2
+// CHECK-NEXT:           [[uint:%\d+]] = OpLoad %uint [[uintPtr]]
+// CHECK-NEXT:           [[bool:%\d+]] = OpINotEqual %bool [[uint]] %uint_0
+// CHECK-NEXT:                           OpStore %f [[bool]]
+    bool    f = frameConstants.boolMat[1][2];
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.boolean.hlsl]: Maybe we should also test swizzling.

File: tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.boolean.hlsl [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+// CHECK-NEXT:        [[boolVec:%\d+]] = OpINotEqual %v3bool [[uintVec]] [[v3uint0]]
+// CHECK-NEXT:                           OpStore %a [[boolVec]]
+    bool3   a = frameConstants.boolVec;
+
+// CHECK:      [[FrameConstants:%\d+]] = OpAccessChain %_ptr_Uniform_FrameConstants %CONSTANTS %int_0
+// CHECK-NEXT:        [[uintPtr:%\d+]] = OpAccessChain %_ptr_Uniform_uint [[FrameConstants]] %int_1 %uint_0
+// CHECK-NEXT:           [[uint:%\d+]] = OpLoad %uint [[uintPtr]]
+// CHECK-NEXT:           [[bool:%\d+]] = OpINotEqual %bool [[uint]] %uint_0
+// CHECK-NEXT:                           OpStore %b [[bool]]
+    bool    b = frameConstants.boolVec[0];
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.boolean.hlsl]: Maybe we should also test swizzling.

File: tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.boolean.hlsl [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+// Swizzle Vector: on temporary value (rvalue)
+// CHECK:       [[uintVec1:%\d+]] = OpLoad %v3uint {{%\d+}}
+// CHECK-NEXT:  [[boolVec1:%\d+]] = OpINotEqual %v3bool [[uintVec1]] [[v3uint0]]
+// CHECK:       [[uintVec2:%\d+]] = OpLoad %v3uint {{%\d+}}
+// CHECK-NEXT:  [[boolVec2:%\d+]] = OpINotEqual %v3bool [[uintVec2]] [[v3uint0]]
+// CHECK-NEXT: [[temporary:%\d+]] = OpLogicalAnd %v3bool [[boolVec1]] [[boolVec2]]
+// CHECK-NEXT:      [[bool:%\d+]] = OpCompositeExtract %bool [[temporary]] 0
+// CHECK-NEXT:                      OpStore %j [[bool]]
+    bool    j = (frameConstants.boolVec && frameConstants.boolVec).x;
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.boolean.hlsl]: Awesome!

File: tools/clang/lib/SPIRV/DeclResultIdMapper.h [hunk 4]
```diff
@@ -300,13 +300,14 @@ class DeclResultIdMapper {
 
   /// \brief Creates a function-scope variable in the current function and
   /// returns its <result-id>.
-  uint32_t createFnVar(const VarDecl *var, llvm::Optional<uint32_t> init);
+  SpirvEvalInfo createFnVar(const VarDecl *var, llvm::Optional<uint32_t> init);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/DeclResultIdMapper.h]: It's not super clear to me. But do we need to change the signature of these `create*Var` methods?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 5]
```diff
@@ -4218,19 +4271,32 @@ SPIRVEmitter::doExtMatrixElementExpr(const ExtMatrixElementExpr *expr) {
...
+    const uint32_t vecType = theBuilder.getVecType(elemTypeId, size);
     valueId = theBuilder.createCompositeConstruct(vecType, elements);
   }
 
+  // Note: Special-case: Booleans have no physical layout, and therefore when
+  // layout is required booleans are represented as unsigned integers.
+  // Therefore, after loading the uint we should convert it boolean.
+  if (elemType->isBooleanType() && layoutRule != LayoutRule::Void) {
+    const auto fromType =
+        size == 1 ? astContext.UnsignedIntTy
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: This code pattern happens a lot. What about a helper function in `TypeTranslator` like
```c++
QualType getScalarOrVectorType(QualType elemType, uint32_t count) {
  return count == 1 ? elemType : astContext.getExtVectorType(elemType);
}
```

Comment (antiagainst) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Meh, minor. It's up to you.

File: tools/clang/lib/SPIRV/DeclResultIdMapper.h (hunk 4 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/DeclResultIdMapper.h]: discussed offline.

**Pull request merged**